src/minisketch/src/test-exhaust.cpp

```diff
@@ -0,0 +1,231 @@
+            minisketch_serialize(state2, nser);
+            // Compare it to the original.
+            CHECK(memcmp(ser, nser, 8) == 0);
+            // Count it.
+            if (num_roots +1 >= (int)ret.size()) ret.resize(num_roots + 2);
+            ret[num_roots + 1]++;
+            minisketch_destroy(state2);
+        } else {
+            if (ret.size() == 0) ret.resize(1);
+            ret[0]++;
+        }
+    }
+    minisketch_destroy(state);
+}
+
+std::vector<uint64_t> TestAll(int bits, int impl, int count, uint32_t threads) {
+    std::vector<std::vector<uint64_t>> outputs;
+    std::vector<std::thread> thread_list;
+    thread_list.reserve(threads);
+    outputs.resize(threads);
+    for (uint32_t i = 0; i < threads; ++i) {
+        thread_list.emplace_back([=,&outputs](){ TestAll(bits, impl, count, i, threads, outputs[i]); });
+    }
+    std::vector<uint64_t> ret;
+    for (uint32_t i = 0; i < threads; ++i) {
+        thread_list[i].join();
+        if (ret.size() < outputs[i].size()) ret.resize(outputs[i].size());
+        for (size_t j = 0; j < outputs[i].size(); ++j) {
+            ret[j] += outputs[i][j];
+        }
+    }
+    if (ret.size()) {
+        for (int i = 1; i <= count + 1; ++i) {
+            CHECK(ret[i] == Combination((uint64_t(1) << bits) - 1, i - 1));
+        }
+    }
+    return ret;
+}
+
+void TestRand(int bits, int impl, int count, int iter) {
+    std::vector<uint64_t> elems(count);
+    std::vector<uint64_t> roots(count + 1);
+    std::random_device rnd;
+    std::uniform_int_distribution<uint64_t> dist(1, bits == 64 ? -1 : ((uint64_t(1) << bits) - 1));
+
+    for (int i = 0; i < iter; ++i) {
+        bool overfill = iter & 1; // Test some cases with overfull sketches that may not decode.
+        minisketch* state = minisketch_create(bits, impl, count);
+        if (!state) return;
+        minisketch* basestate = minisketch_create(bits, 0, count);
+        for (int j = 0; j < count + 3*overfill; ++j) {
+            uint64_t r = dist(rnd);
+            if (!overfill) elems[j] = r;
+            minisketch_add_uint64(state, r);
+            minisketch_add_uint64(basestate, r);
+        }
+        roots.assign(count + 1, 0);
+        std::vector<unsigned char> data, basedata;
+        basedata.resize(((count + 1) * bits + 7) / 8);
+        data.resize(((count + 1) * bits + 7) / 8);
+        minisketch_serialize(basestate, basedata.data());
+        minisketch_serialize(state, data.data());
+        CHECK(data == basedata);
+        minisketch_deserialize(state, basedata.data());
+        int num_roots = minisketch_decode(state, count + 1, roots.data());
+        CHECK(overfill || num_roots >= 0);
+        CHECK(num_roots < 1 || minisketch_decode(state, num_roots - 1, roots.data()) == -1); // Decoding with a too-low maximum should fail.
+        if (!overfill) {
+            std::sort(roots.begin(), roots.begin() + num_roots);
+//            fprintf(stderr, "Solut: ");
```

Comment:
wouldn't it be better to have this guarded by an ifdef?

Reply:
I think github.com/sipa/minisketch master branch is a better place for these comments :)